libcloudforensics/aws.py

```diff
@@ -0,0 +1,695 @@
+
+  def ClientApi(self, service, region=None):
+    """Create an AWS client object.
+
+    Args:
+      service (str): The AWS service to use.
+      region (str): Optional. The region is which to create new resources in.
+
+    Returns:
+      boto3.Session.Client: An AWS EC2 client object.
+    """
+    if region:
+      return boto3.session.Session().client(
+          service_name=service, region_name=region
+      )
+    return boto3.session.Session().client(
+        service_name=service, region_name=self.default_region
+    )
+
+  def ResourceApi(self, service, region=None):
+    """Create an AWS resource object.
+
+    Args:
+      service (str): The AWS service to use.
+      region (str): Optional. The region is which to create new resources in.
+
+    Returns:
+      boto3.Session.Resource: An AWS EC2 resource object.
+    """
+    if region:
+      return boto3.session.Session().resource(
+          service_name=service, region_name=region
+      )
+    return boto3.session.Session().resource(
+        service_name=service, region_name=self.default_region
+    )
+
+  def ListInstances(self, region=None, filters=None):
+    """List instances of an AWS account.
+
+    Args:
+      region (str): Optional. The region from which to list instances.
+      filters (list(dict)): Optional. Filters for the query.
+
+    Returns:
+      dict: Dictionary with name and metadata for each instance.
+
+    Raises:
+      RuntimeError: If instances can't be listed.
+
+    Example usage:
+      ListInstances(region='us-east-1', filters=[
+          dict(Name='instance-id', Values=['some-instance-id'])])
+    """
+    if not filters:
+      filters = []
+
+    instances = {}
+    next_token = None
+    client = self.ClientApi(EC2_SERVICE, region=region)
+
+    while True:
+      try:
+        if next_token:
+          response = client.describe_instances(
+              Filters=filters, NextToken=next_token)
+        else:
+          response = client.describe_instances(Filters=filters)
+      except client.exceptions.ClientError as exception:
+        raise RuntimeError('Error: could not retrieve instances: {0:s}'.format(
+            str(exception)))
+
+      for reservation in response['Reservations']:
+        for instance in reservation['Instances']:
+          # Terminated instances are filtered out. If
+          # reservation['Instances'] contains any entry, then the
+          # instance's state is expected to be present in the API's response.
+          if instance['State']['Name'] == 'terminated':
+            continue
+
+          zone = instance['Placement']['AvailabilityZone']
+          instance_info = {
+              'region': zone[:-1],
+              'zone': zone
+          }
+
+          for tag in instance.get('Tags', []):
+            if tag.get('Key') == 'Name':
+              instance_info['name'] = tag.get('Value')
+              break
+
+          instances[instance['InstanceId']] = instance_info
```

Comment:
Why not use AWSInstance as values in the dict "instances"?

Reply:
The original thought was to be as consistent as possible with the methods' responses that `gcp.py` defined. Should we adapt both, or leave it as is?

Reply:
Oh I see.
Maybe add a TODO or a new issue to consider having some kind of generic Instance and Disk interface/object could be nice, since all of these basically have names, zones, regions, etc

Reply:
https://github.com/google/cloud-forensics-utils/issues/32